### PR TITLE
Context parallelism for Megatron core models

### PR DESCRIPTION
**Pull request by yeyu-nvidia**
## What does this PR do?

New feature

**Overview:** 
This PR implements the context manager which injects attn_mask as attn_bias to TEDotProductAttention so that we can enable EAGLE training with arbitrary mask.

## Usage
set CP>1 in https://github.com/NVIDIA/Megatron-LM/blob/main/examples/post_training/modelopt/finetune.sh

```python
# Add a code snippet demonstrating how to use this
```

## Testing
Tested on DSR1 Llama 8B.
CP1->CP2
38854MB->28050MB
MTbench AL 2.26->2.31

## Before your PR is "*Ready for review*"

- **Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/Model-Optimizer/blob/main/CONTRIBUTING.md)** and your commits are signed.
- **Is this change backward compatible?**: Yes/No 
- **Did you write any new necessary tests?**: Yes/No
- **Did you add or update any necessary documentation?**: Yes/No
- **Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?**: Yes/No 

## New Features
* Context parallelism support added for Eagle speculative decoding with HuggingFace and Megatron Core models.
* Model checkpoint loading enhanced to enable remote code execution capabilities when required.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: modelopt/torch/speculative/plugins/megatron_eagle.py [hunk 1]
```diff
@@ -838,16 +847,41 @@ def _get_eagle_module_inputs(
...
+            padded_input_ids, group=get_context_parallel_group()
+        )
+        # [s,b] -> [b,s]
+        padded_input_ids = padded_input_ids.transpose(0, 1).contiguous()
+
         attn_mask = attention_mask.clone().detach()
-        attn_mask[:, :, :-1, :-1] = attention_mask[:, :, 1:, 1:]
+        # [b, 1, sq, sk] -> [sq, 1, b, sk]
+        attn_mask = attn_mask.transpose(0, 2).contiguous()
+        attn_mask = gather_from_sequence_parallel_region(
```
Comment (ChenhanYu) [inline modelopt/torch/speculative/plugins/megatron_eagle.py:885]: I assume here that here attn_mask has shape `[sq*cp, 1, b, sk]`. My understanding is that `sk=ttt_step*sq*cp`? So the attn_mask is O(ttt_step*b*sq^2*cp^2)?

File: modelopt/torch/speculative/plugins/megatron_eagle.py [hunk 2]
```diff
@@ -860,9 +894,17 @@ def _get_eagle_module_inputs(
...
             position_ids=eagle_inputs["position_ids"],
         )
+
         eagle_inputs["hidden_states"] = hidden_states
 
-        eagle_inputs["attention_mask"] = set_multi_step_attention_mask(attn_mask, ttt_step)
+        attn_mask = set_multi_step_attention_mask(attn_mask, ttt_step)
+        # [b, 1, sq, sk] -> [sq, 1, b, sk]
+        attn_mask = attn_mask.transpose(0, 2).contiguous()
+        attn_mask = scatter_to_sequence_parallel_region(
```
Comment (ChenhanYu) [inline modelopt/torch/speculative/plugins/megatron_eagle.py:909]: I assume scatter is from `[cp*sq, 1, b, sk]` back to `[sq, 1, b, sk]`?

File: modelopt/torch/speculative/plugins/megatron_eagle.py [hunk 3]
```diff
@@ -1330,3 +1378,87 @@ def get_ground_truth(self, input_ids, osl):
             if input_id[0, 0] == self.end_token:
                 break
         return input_ids
+
+
+@contextmanager
+def te_dot_product_attention_with_cp(attention_mask: torch.Tensor, num_attention_heads: int):
```
Comment (ChenhanYu) [inline modelopt/torch/speculative/plugins/megatron_eagle.py:1387]: I think I need a walkthrough here.

File: modelopt/torch/speculative/plugins/megatron_eagle.py (hunk 1 above)
Comment (yeyu-nvidia) [reply modelopt/torch/speculative/plugins/megatron_eagle.py:885]: sk=sq*cp. ttt_step will be introduced later at set_multi_step_attention_mask

File: modelopt/torch/speculative/plugins/megatron_eagle.py (hunk 2 above)
Comment (yeyu-nvidia) [reply modelopt/torch/speculative/plugins/megatron_eagle.py:909]: yes

**Review by ChenhanYu — APPROVED** [review]
